### DISCUSSION TITLE
HTTP POST request to /api throw "Error: Unexpected end of form" at Multipart._final

### DISCUSSION THREAD

**Discussion by immotus** [payloadcms/payload#5969] [Q&A]
### Describe the Bug

When Payload is deployed into GCP Cloud Run as a Firebase Function, the `/api/users/login` POST request, and by extension the login form on the Admin UI, fail with the following error:

```json
{
	"errors": [
		{
			"message": "Unexpected end of form"
		}
	],
	"stack": "Error: Unexpected end of form
    at Multipart._final (/workspace/node_modules/busboy/lib/types/multipart.js:588:17)
    at prefinish (node:internal/streams/writable:907:14)
    at finishMaybe (node:internal/streams/writable:921:5)
    at Writable.end (node:internal/streams/writable:836:5)
    at onend (node:internal/streams/readable:946:10)
    at process.processTicksAndRejections (node:internal/process/task_queues:77:11)"
}
```

Other API methods seem to work fine, e.g., `/api/users/init` returns:

```json
{
	"initialized": true
}
```

Two days of searching resulted in many reports of the same issue, which seems to be linked to the `express-fileupload` and/or `busboy` packages. For example, see https://github.com/expressjs/multer/issues/1144. None of the recommendations worked, either.

Locally, everything works fine.

### To Reproduce

Deploy the Payload e-commerce template as a Cloud Run function into GCP and connect to a GCP-hosted Mongo Atlas cluster.

Dependencies:

```json
"@payloadcms/bundler-webpack": "^1.0.6",
"@payloadcms/db-mongodb": "^1.4.4",
"@payloadcms/richtext-slate": "^1.5.0",
"payload": "^2.12.1",
```

Cloud Run environment:

node: 20
Region: us-central1
Memory allocated: 256 MiB
CPU: 1

### Payload Version

2.12.1

### Adapters and Plugins

bundler-webpack, db-mongodb, richtext-slate

**Comment by immotus** [comment]
It appears that it might be linked to the Firebase Functions (i.e. Cloud Run) environment itself:

- https://stackoverflow.com/questions/75922301/cant-submit-form-when-using-firebase-functions
- https://stackoverflow.com/questions/78319321/busboy-not-working-with-firebase-firestorage

**Comment by jmikrut** [comment]
Hey @PavelGolodoniuc — good find. I am not aware of anyone in our community successfully deploying to Firebase Functions, but it sounds like there might not be much we can do to fix this.

However, I do think that from your Stack Overflow responses, it seems like Multer is incompatible with Firebase Functions, and Busboy is not. 

In 3.x, we are removing our reliance on Multer and just using Busboy itself. So your issue may actually be fixed on 3.x.

In the meantime, I will convert this to a discussion so others can find it. Let us know how you get on!

**Reply by immotus** [reply]
@jmikrut — well, now you have! The community is growing 🙂

After a little bit of fiddling, I eventually succeeded with the deployment into GCP Cloud Run, although I circumvented the Firebase Functions altogether. I ended up packaging up the Payload application in a Docker container, pushing it into the GCP Artifacts Repository, and then directly deploying it into Cloud Run. Everything works fine and scales appropriately. The Firebase Functions SDK supposedly does the same thing, but apparently, it applies some black magic when it containerizes the node.js application.

Looking forward to the version 3 release!
